src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs

```diff
@@ -0,0 +1,273 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.Json;
+
+namespace NuGet.ProjectModel
+{
+    /// <summary>
+    /// This struct is used to read over a memeory stream in parts, in order to avoid reading the entire stream into memory.
+    /// It functions as a wrapper around <see cref="Utf8JsonStreamReader"/>, while maintaining a stream and a buffer to read from.
+    /// </summary>
+    internal ref struct Utf8JsonStreamReader
+    {
+        private static readonly char[] DelimitedStringDelimiters = [' ', ','];
+        private static readonly byte[] Utf8Bom = [0xEF, 0xBB, 0xBF];
+
+        private const int BufferSizeDefault = 16 * 1024;
+        private const int MinBufferSize = 1024;
+        private Utf8JsonReader _reader;
+        // The buffer is used to read from the stream in chunks.
+        private byte[] _buffer;
+        private bool _disposed;
+        private ArrayPool<byte> _bufferPool;
+        private int _bufferUsed = 0;
+
+        internal Utf8JsonStreamReader(Stream stream, int bufferSize = BufferSizeDefault, ArrayPool<byte> arrayPool = null)
+        {
+            if (stream is null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            if (bufferSize < MinBufferSize)
+            {
+                throw new ArgumentException($"Buffer size must be at least {MinBufferSize} bytes", nameof(bufferSize));
+            }
+
+            _bufferPool = arrayPool ?? ArrayPool<byte>.Shared;
+            _buffer = _bufferPool.Rent(bufferSize);
+            _disposed = false;
+            Stream = stream;
+            Stream.Read(_buffer, 0, 3);
+            if (!Utf8Bom.AsSpan().SequenceEqual(_buffer.AsSpan(0, 3)))
+            {
+                _bufferUsed = 3;
+            }
+
+            var iniialJsonReaderState = new JsonReaderState(new JsonReaderOptions
+            {
+                AllowTrailingCommas = true,
+                CommentHandling = JsonCommentHandling.Skip,
+            });
+
+            ReadStreamIntoBuffer(iniialJsonReaderState);
+            _reader.Read();
+        }
+
+        private Stream Stream { get; set; }
```

Comment:
nitpick (sorry if I didn't notice this earlier):  private auto-properties don't really have a lot of value, and isn't part of common .NET conventions. Maybe the JIT compiler optimizes it to the same perf as a field, but I think it's clearer to just use a field rather than a private auto-prop.

Reply:
This one is super annoying, since Utf8JsonStreamReader is disposable all of the fields it has must also be disposed in the Dispose() method. I don't want to dispose of the stream though because it's passed in and that feels like it could be an issue.

Reply:
> Utf8JsonStreamReader is disposable all of the fields 

How did you come to this conclusion? It should only dispose fields it owns. The auto property versus field doesn't change this anyway as they both end up with a field of type "Stream".

Reply:
I think there's a built-in roslyn analyzer when a field that implements `IDisposable` isn't disposed anywhere in the class (maybe it even validates that it must be disposed from the `Dispose()` method, I'm not sure).

However, we can suppress that analyzer in this class, if that's correct.  Plus add XMLDoc to the class, maybe the constructor too, to make it clear that disposing this struct won't dispose the stream.

Also consider that classes like `StreamReader` have overloads for `bool leaveOpen`. I'm not saying we have to copy that pattern, but if you think it makes sense, then it's an option. But maybe we never want to auto-dispose the underlying stream for this type 🤷